libs/stokesian_dynamics/sd.hpp

```diff
@@ -0,0 +1,1478 @@
+#include <cstdio>
+#include <vector>
+
+#include "thrust_wrapper.hpp"
+#include "device_matrix.hpp"
+#include "multi_array.hpp"
+
+#ifdef __CUDACC__
+#include <curand_kernel.h>
+#else
+#include <Random123/philox.h>
+
+typedef r123::Philox2x64 RNG;
+#endif
+
+
+#ifdef __CUDACC__
+#define DEVICE_FUNC __host__ __device__
+#define KERNEL_ABORT asm("trap;")
+#else
+#define DEVICE_FUNC
+#define KERNEL_ABORT abort()
+#endif
+
+/** \file
+ *  This file contains computations required for Stokesioan Dynamics, namely
+ *  the particle's translational and angular velocities are computed from the 
+ *  particle's positions and radii and the forces and torques that are acting
+ *  on the particles.
+ *
+ *  In its current implementation, the neccessary subroutines are implemented
+ *  as functors. The main functor is called solver, which wraps up the
+ *  functionality of all the other functors in this file. For details, please
+ *  see the description of the solver functor.
+ *
+ *  For details on the thermalization, please see the description of the
+ *  thermalizer functor.
+ *
+ *  The general Stokesian Dynamics method also allows to incorporate one
+ *  stress tensor per particle which gives the stress that is imposed on a
+ *  particle by the surrounding fluid (such as to deform the particle).
+ *  However, ESPResSo makes no use of that, leaving only forces and torques
+ *  to be passed to the SD routine.
+ *
+ *  To understand the Stokesian Dynamics method, it is recommended to read
+ *  Durlofski et al. (1987), because it describes the method that is
+ *  implemented here, apart from the thermalization.
+ *  (F-T-version or F-T-S-version?)
+ *      "Dynamic simulation of hydrodynamically interacting suspensions"
+ *      https://core.ac.uk/download/pdf/4895234.pdf
+ *
+ *  All references to formulae in this file refer to the aforementioned paper,
+ *  unless otherwise noted.
```

Comment:
please change "refer to the aforementioned paper" to `refer to @cite paper`: it's easy to accidentally change the order of papers when updating a docstring